Step3 -  구간 추가 기능

yulmucha opened the pull request:
안녕하세요.

- 구간 A-C 사이에 역 B를 추가하는 방법으로는 구간 A-B 추가뿐만 아니라 구간 B-C를 추가도 가능합니다.
- 구간 목록이 비어 있을 때 구간, 예를 들어 A-C를 추가하면 구간 null-A가 먼저 자동으로 추가되고 나서 구간 A-C가 추가되도록 했습니다.
- 위 두 항목에 의해, 새로운 역을 상행 종점으로 등록하는 경우는 사실 구간 null-상행종점 사이에 새로운 역을 추가하는 것이 됩니다.
- validation마다 매번 section list를 도는 게 좀 찝찝하긴 하지만 많아 봐야 100개 정도 들어 있을 테니 문제 없을 것 같긴 합니다.

감사합니다.

Commit:
refactor(LineAcceptanceTest, StationAcceptanceTest): use DTO as request body

Commit:
refactor(StationService, StationRepository): move the check on name uniqueness to service

Commit:
refactor(LineAcceptanceTest, StationAcceptanceTest): use location header instead of the id of a created resource

Commit:
feat(Section, Sections, Line, LineService, LineResponse, truncate.sql): add Section entity and map it to Line and Station

Commit:
feat(Section, Sections, Line, SectionService, LineController, LineResponse): add a station between stations and show stations of a line

Commit:
test(SectionAcceptanceTest): add cases that last station is replaced

Commit:
feat(Sections, Section): validate Section when added to prevent duplication

Commit:
feat(Sections, Section): validate Section's distance when added

Commit:
feat(Sections, Section): validate Section's stations when added

Commit:
test(SectionAcceptanceTest): add cases of adding various sections and a duplicate section

Commit:
feat(Section, SectionAcceptanceTest): reduce distance when station is inserted

Commit:
docs(README.md): add requirements

Commit:
refactor(SectionAcceptanceTest): change function parameter

Commit:
refactor(Line, Sections): move responsibility to add first section to addSection()

Commit:
test(SectionsTest, Sections, Section): add unit test for Sections

Commit:
refactor(Sections): extract methods

seokhongkim commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -0,0 +1,114 @@
+                });
+    }
+
+    private void validateDistance(Section newSection) {
```
이 밸리데이션은 역을 추가할 구간을 먼저 확인한 후, 해당 구간에 대해서만 밸리데이션을 수행해도 괜찮을 것 같네요.

yulmucha replied:
그런 식으로 하니 훨씬 간단해졌네요! 🥳

seokhongkim commented on src/main/java/nextstep/subway/domain/Sections.java:
```diff
@@ -0,0 +1,114 @@
+    @OneToMany(mappedBy = "line", cascade = CascadeType.ALL, orphanRemoval = true)
+    private List<Section> list = new ArrayList<>();
+
+    public void add(Section newSection) {
```
전율님은 메소드 배치 순서를 어떤 순서로 작성하시나요?

yulmucha replied:
기본적으로는 public 메서드 다음에 private 메서드가 오도록 배치합니다.
그리고 public과 private 내에서는 고정된 기준은 없지만,
보통 get 성격의 메서드 또는 중요하다 싶은 메서드를 위로 올리거나, 비슷한 성격의 메서드끼리 이웃하게 배치하거나, 코드 상 출현 순으로 배치하거나, 추가된 시간 순으로 배치하거나 합니다.
그런데 다시 보니까 좀 섞여있었군요 😅

seokhongkim replied:
여러 기준이 있을 것 같은데, 코드를 보고 이해하기 쉬운 순서가 어떤 순서일지 고민해보셔도 좋을 것 같습니다. :)
클린 코드 책을 참고해보셔도 좋고, 이와 관련해서 슬랙 채널에 올라온 내용을 참고해보셔도 도움이 될 것 같아서 링크 공유드립니다!
> https://nextstep-members.slack.com/archives/C0253RP321J/p1654409524938299

seokhongkim commented on src/main/java/nextstep/subway/application/StationService.java:
```diff
@@ -20,8 +21,13 @@ public StationService(StationRepository stationRepository) {
-    public StationResponse saveStation(StationRequest stationRequest) {
-        Station persistStation = stationRepository.saveOne(stationRequest.toStation());
+    public StationResponse saveStation(StationRequest request) {
+        stationRepository.findByName(request.getName())
```
꼼꼼한 예외 처리 좋습니다. 👍 

나중에 실무 코드에서는 DB의 유니크 제약조건으로 발생하는 예외가 있을 수 있으므로 이에 대한 예외 처리도 같이되면 좋을 것 같습니다.

seokhongkim commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -0,0 +1,113 @@
+    }
+
+    public boolean isFirstSection() {
+        return Objects.isNull(upStation);
```
상행 종점에 대해서는 `upStation` 이 null 을 갖도록 구현하셨네요.
하행 종점은 다른 구간과 동일하게 구현하셨는데, 둘을 다르게 구현하신 이유가 있을까요?

yulmucha replied:
처음엔 구간(`Section`)이 아니라 전 역에 대한 정보도 갖고 있는 지하철 역(`LineStation`)으로 개념을 잡고(이 경우 이전 역은 `null`일 수 있지만 현재 역은 `null`일 수 없음) 구현하다가 이름을 `Section`으로 바꾸었습니다.
하행 마지막 구간도 마찬가지로 `downStation`이 `null`을 갖도록 하면 하행 종점 찾기도 쉽고 더 간편하지 않을까 싶습니다.

yulmucha replied:
`downStation`이 `null`인 마지막 구간도 추가하도록 수정했습니다.
그러면서 책임도 좀 이동시켰더니 일관성이 생긴 것 같네요. `SectionsTest`의 `setup`에서 `upStation`이 `null`인 구간을 직접 넣어주던 부분도 사라졌습니다. 😌

seokhongkim commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -0,0 +1,113 @@
+        this.distance -= newSection.distance;
+    }
+
+    public void updateDownStationToUpStationOf(Section newSection) {
```
상행 종점을 추가하는 경우, 상행 종점의 distance 는 의미 없는 값이 되면서 의도치 않은 값을 갖게 될 것 같네요!

yulmucha replied:
종점 구간(null-상행종점, 하행종점-null)은 `distance`가 항상 `0`이 되도록 수정했습니다.

seokhongkim commented on src/test/java/nextstep/subway/section/SectionAcceptanceTest.java:
```diff
@@ -0,0 +1,275 @@
+    public void setUp() {
+        RestAssured.port = port;
+
+        청량리역 = StationAcceptanceTest.createStation(new StationRequest("청량리역")).as(StationResponse.class);
```
실제 노선을 가지고 테스트를 작성해주셔서 이해하기 쉬웠습니다. 👍

seokhongkim commented on src/test/java/nextstep/subway/section/SectionsTest.java:
```diff
@@ -0,0 +1,104 @@
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+public class SectionsTest {
```
주요 비즈니스에 대한 꼼꼼한 유닛 테스트 좋습니다. 👍

seokhongkim reviewed (changes requested):
전율님, ATDD 3단계 미션 잘 구현해주셨습니다. 👍 
인수 테스트와 유닛 테스트를 잘 작성해주시고, 커밋을 작은 단위로 나눠서 해주신 점이 좋았습니다.

코드에 코멘트를 조금 남겼습니다. 확인 부탁드립니다. 🙏

Commit:
feat(StationService, ControllerExceptionHandler): handler exception from save

Commit:
refactor(Line, Sections, Section, SectionTest): add the section having null down station as the last section
- move responsibility to add first and last section to Sections
- move responsibility to create first and last section to Section
- modify validations of Sections

seokhongkim commented on src/main/java/nextstep/subway/application/StationService.java:
```diff
@@ -27,8 +28,12 @@ public StationResponse saveStation(StationRequest request) {
+            Station persistStation = stationRepository.save(request.toStation());
+            return StationResponse.of(persistStation);
+        } catch (DataIntegrityViolationException e) {
+            throw new DataIntegrityViolationException("유니크 제약 조건 위반: " + request.getName());
```
예외를 처리해주신 덕분에 26~29번째 라인에 있는 사전 검증은 없어도 괜찮을 것 같네요! 👍

seokhongkim commented on src/main/java/nextstep/subway/domain/Section.java:
```diff
@@ -50,39 +52,47 @@ public boolean hasSameDownStationAs(Section newSection) {
     public void updateUpStationToDownStationOf(Section newSection) {
         this.upStation = newSection.downStation;
-        this.distance -= newSection.distance;
+        this.distance = Math.max(this.distance - newSection.distance, 0);
```
`Math.max` 를 이용하여 음수 거리가 될 수 없도록 처리하신 방법도 좋네요.

저는 예외 처리를 하면 나중에 코드를 볼 때 히스토리를 모르면 이해하기 어려운 경우가 종종 있어서, 아래와 같은 방법을 사용합니다.
1. 예외 처리를 코드만 보고 이해할 수 있도록 변경한다. (e.g., 종점이면 거리를 0으로 설정한다.)
2. 예외 처리를 한 이유에 대해 코드에 주석을 남긴다. 로직이 아닌, 왜 예외 처리를 한 것인지에 대한 설명을 위주로 작성한다.

seokhongkim reviewed (approved):
전율님, 3단계 미션 피드백 잘 반영해주셨네요! 👍 
3단계 PR은 이만 머지하도록 하겠습니다.
코멘트 남긴 내용은 다음 단계 미션 진행하실 때 참고하셔도 좋을 것 같습니다.

4단계 미션에서 뵙겠습니다. 🙏
